thingif/src/main/java/com/kii/thingif/ThingIFAPI.java

```diff
@@ -1002,30 +1003,26 @@ private Trigger postNewTriggerWithForm(
         if (predicate == null) {
             throw new IllegalArgumentException("predicate is null.");
         }
-        //TODO: // FIXME: 2017/01/23
-        return null;
-//        JSONObject requestBody = options != null ?
-//                JsonUtils.newJson(GsonRepository.gson().toJson(options)) :
-//                new JSONObject();
-//        try {
-//            requestBody.put("triggersWhat", TriggersWhat.COMMAND.name());
-//            requestBody.put("predicate", JsonUtils.newJson(
-//                        GsonRepository.gson().toJson(predicate)));
-//            //TODO: // FIXME: 12/15/16 fix the parse code
-//            JSONObject command = JsonUtils.newJson(
-//                GsonRepository.gson(
-//                    ).toJson(form));
-//            command.put("issuer", this.owner.getTypedID());
-//            if (form.getTargetID() == null) {
-//                command.put("target", this.target.getTypedID().toString());
-//            }
-//            requestBody.put("command", command);
-//        } catch (JSONException e) {
-//            // Won't happen.
-//            // TODO: remove this after test finished.
-//            throw new RuntimeException(e);
-//        }
-//        return postNewTrigger(requestBody);
+        JSONObject requestBody = options != null ?
+                JsonUtils.newJson(this.gson.toJson(options)) :
+                new JSONObject();
+        try {
+            requestBody.put("triggersWhat", TriggersWhat.COMMAND.name());
+            requestBody.put("predicate", JsonUtils.newJson(
+                        this.gson.toJson(predicate, Predicate.class)));
+            JSONObject command = JsonUtils.newJson(
+                this.gson.toJson(form));
+            command.put("issuer", this.owner.getTypedID());
+            if (form.getTargetID() == null) {
+                command.put("target", this.target.getTypedID().toString());
```

Comment:
I think

```java
if (!command.has("target")) {
```

is better than

```java
if (form.getTargetID() == null)
```

because we need to put `target` field if `command` does not have the field. I think checking `command` is more clear.

But this might be just my liking. If you do not like my suggestion. Please ignore this:)